src/ristretto_point.rs

```diff
@@ -0,0 +1,218 @@
+use crate::util::nonzero_gadget;
+use bulletproofs::r1cs::{
+    ConstraintSystem, LinearCombination, R1CSError, RandomizedConstraintSystem, Variable,
+};
+use curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar};
+use zerocaf::ristretto::RistrettoPoint as SonnyRistrettoPoint;
+use zerocaf::scalar::Scalar as SonnyScalar;
+use zerocaf::traits::ops::Double;
+
+#[derive(Clone)]
+// Represents a Sonny Edwards Point using Twisted Edwards Extended Coordinates
+pub struct SonnyRistrettoPointGadget {
+    pub X: LinearCombination,
+    pub Y: LinearCombination,
+    pub Z: LinearCombination,
+    pub T: LinearCombination,
+}
+
+impl SonnyRistrettoPointGadget {
+    /// Builds a `SonnyRistrettoPointGadget` from a `SonnyRistrettoPoint` adding a constrain
+    /// that checks that the point relies on the curve and another one checking that
+    /// it is indeed a RistrettoPoint.
+    pub fn from_point(point: SonnyRistrettoPoint, cs: &mut dyn ConstraintSystem) -> Self {
+        let gadget_p = SonnyRistrettoPointGadget {
+            X: Scalar::from_bytes_mod_order(point.0.X.to_bytes()).into(),
+            Y: Scalar::from_bytes_mod_order(point.0.Y.to_bytes()).into(),
+            Z: Scalar::from_bytes_mod_order(point.0.Z.to_bytes()).into(),
+            T: Scalar::from_bytes_mod_order(point.0.T.to_bytes()).into(),
+        };
+        gadget_p.ristretto_gadget(cs, Some(point));
+        gadget_p
+    }
+
+    pub fn from_LCs(lcs: Vec<LinearCombination>, cs: &mut ConstraintSystem) -> Self {
+        assert!(lcs.len() == 4);
+        let gadget = SonnyRistrettoPointGadget {
+            X: lcs[0].clone(),
+            Y: lcs[1].clone(),
+            Z: lcs[2].clone(),
+            T: lcs[3].clone(),
+        };
+
+        gadget.ristretto_gadget(cs, None);
+        gadget
+    }
+
+    /// Adds constrains to validate only points that lie on the prime sub-group and excludes the others
+    /// that lie on smaller order groups with order (2, 4 and 8).
+    /// It also adds constrains that validate only points that satisfy the Sonnycurve equation.
+    pub fn ristretto_gadget(
+        &self,
+        cs: &mut dyn ConstraintSystem,
+        point_assign: Option<SonnyRistrettoPoint>,
+    ) {
+        // XXX: Here we should check that the point relies on the curve.
```

Comment:
I'm not sure that's possible on zero-knowledge. We are talking about reassuring the curve consistency outside the circuit?

Reply:
Inside.

Indeed we need to check inside the R1CS that the points used on the input are really RistrettoPoints. That's why this gadget exists.